[PGO] Fix sandboxing issues and simplify collection

emw-apple opened the pull request:
#### 6fc61a30ebc7388f132ccd447db31ed6c04612cc
<pre>
[PGO] Fix sandboxing issues and simplify collection
<a href="https://bugs.webkit.org/show_bug.cgi?id=261014">https://bugs.webkit.org/show_bug.cgi?id=261014</a>
rdar://113870737

Reviewed by Justin Michaud and Alexey Proskuryakov.

Make a number of simplifications to our PGO instrumentation logic
to fix instances where profile collection would fail silently due to
sandboxing:

Instead of initializing and writing out instrumented profiles ourselves,
use LLVM&apos;s built-in logic. We bake in a default profile path into
instrumented binaries, which tells the instrumentation machinery to
write profiles to /private/tmp/WebKitPGO. This is done by declaring
__llvm_profile_filename in each instrumented binary, rather than using
the -fprofile-generate=&lt;name&gt; compiler argument, because there is no
Xcode build setting to easily set it, see rdar://114792050. At process
launch, profiling begins via a static initializer compiled in to each
instrumented binary from LLVM.

Auto-initalization works great for starting PGO collection in a
predictably, across all WebKit processes, but termination requires extra
care. On Darwin-based platforms, we kill WebKit XPC processes via _exit,
without running atexit handlers, which is normally when profiles would
be written. To accomodate this, run profile collection in &quot;continuous
mode&quot;, denoted by a %c in the profile string. LLVM mmaps the profile
file and records function calls directly in the mapped file, with no
writeback step needed.

Add sandbox rules to permit writing to the WebKitPGO directory when
built for instrumentation. Teach the benchmark runner to copy from
WebKitPGO when Safari exits, instead of parsing the system log to find
written profiles.

* Source/JavaScriptCore/runtime/InitializeThreading.cpp: Declare default
  profile name for JSC.
* Source/WTF/WTF.xcodeproj/project.pbxproj: Remove GenerateProfiles.h.
* Source/WTF/wtf/CMakeLists.txt: Remove GenerateProfiles.h.
* Source/WTF/wtf/GenerateProfiles.h: Removed.
* Source/WebCore/bindings/js/ScriptController.cpp: Declare default
  profile name for WebCore.
* Source/WebKit/Configurations/BaseXPCService.xcconfig: XPC executables
  (i.e. the binary that contains XPCServiceMain) were being
  instrumented, but the profiles were being discarded due to sandboxing.
  We can safely turn off instrumentation -- the only binaries we care
  about are the executables in the WebKit framework stack.
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in: Allow
  access to the WebKitPGO directory.
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in: Ditto.
* Source/WebKit/Shared/Cocoa/WebKit2InitializeCocoa.mm: Declare default
  profile name for WebKit.
* Source/WebKit/Shared/WebKit2Initialize.cpp: Remove spurious include.
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in: Ditto above.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py: When
  collecting profiles, each invocation will populate /tmp/WebKitPGO.
  Ensure the directory is empty before launching the browser, and copy
  its contents to the diagnostic directory upon exit.
(BenchmarkRunner._run_benchmark):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(run_benchmark_plan): Use a fixed path for writing profiles to.
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner._get_result): Remove the logic that listened
  to the system log for profiling messages and moved profiles out ad hoc.
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig: It static-links
  libWebKitPlatform.a, so when that library is compiled with profiling
  it needs it too. This has likely always been broken, and not been an
  issue since we never attempt to build for profiling with tools.

Canonical link: <a href="https://commits.webkit.org/267741@main">https://commits.webkit.org/267741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9430855971b2e1874f479a23834d954b4fa97feb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18790 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15913 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17469 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18133 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17553 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14737 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19605 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/17141 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15431 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22138 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14671 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15780 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19917 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16227 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13725 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18566 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15355 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4328 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4181 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19719 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19791 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16028 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4186 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

Do the continuous profiles actually work? I seem to recall we didn't use them originally because they didn't work properly.

Also, does this work on iOS now?

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_safari_driver.py:
```diff
@@ -1,8 +1,9 @@
 import os
-import sys
 import re
+import shutil
```
unused import?

emw-apple replied:
Indeed, thanks!

aproskuryakov replied:
Wait, but I see `shutil.copytree` below?

emw-apple replied:
…in `benchmark_runner.py`, not this file. (This import is in fact because I was originally making the `copytree` call from the driver)

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -176,6 +180,8 @@ def _run_benchmark(self, count, web_root):
                 finally:
                     self._browser_driver.restore_env()
+                if self._pgo_profile_output_dir:
+                    shutil.copytree(self._pgo_profile_output_dir, self._diagnose_dir, dirs_exist_ok=True)
```
Do we want to assert the existing of `self._pgo_profile_output_dir`?

emw-apple replied:
I don't think so. It's created above (L157), so if it does not exist at this point, something has gone horribly wrong and it would be appropriate to crash with a `FileNotFoundError`.

emw-apple committed:
[PGO] Fix sandboxing issues and simplify collection
https://bugs.webkit.org/show_bug.cgi?id=261014
rdar://113870737

Reviewed by Justin Michaud and Alexey Proskuryakov.

Make a number of simplifications to our PGO instrumentation logic
to fix instances where profile collection would fail silently due to
sandboxing:

Instead of initializing and writing out instrumented profiles ourselves,
use LLVM's built-in logic. We bake in a default profile path into
instrumented binaries, which tells the instrumentation machinery to
write profiles to /private/tmp/WebKitPGO. This is done by declaring
__llvm_profile_filename in each instrumented binary, rather than using
the -fprofile-generate=<name> compiler argument, because there is no
Xcode build setting to easily set it, see rdar://114792050. At process
launch, profiling begins via a static initializer compiled in to each
instrumented binary from LLVM.

Auto-initalization works great for starting PGO collection in a
predictably, across all WebKit processes, but termination requires extra
care. On Darwin-based platforms, we kill WebKit XPC processes via _exit,
without running atexit handlers, which is normally when profiles would
be written. To accomodate this, run profile collection in "continuous
mode", denoted by a %c in the profile string. LLVM mmaps the profile
file and records function calls directly in the mapped file, with no
writeback step needed.

Add sandbox rules to permit writing to the WebKitPGO directory when
built for instrumentation. Teach the benchmark runner to copy from
WebKitPGO when Safari exits, instead of parsing the system log to find
written profiles.

* Source/JavaScriptCore/runtime/InitializeThreading.cpp: Declare default
  profile name for JSC.
* Source/WTF/WTF.xcodeproj/project.pbxproj: Remove GenerateProfiles.h.
* Source/WTF/wtf/CMakeLists.txt: Remove GenerateProfiles.h.
* Source/WTF/wtf/GenerateProfiles.h: Removed.
* Source/WebCore/bindings/js/ScriptController.cpp: Declare default
  profile name for WebCore.
* Source/WebKit/Configurations/BaseXPCService.xcconfig: XPC executables
  (i.e. the binary that contains XPCServiceMain) were being
  instrumented, but the profiles were being discarded due to sandboxing.
  We can safely turn off instrumentation -- the only binaries we care
  about are the executables in the WebKit framework stack.
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in: Allow
  access to the WebKitPGO directory.
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in: Ditto.
* Source/WebKit/Shared/Cocoa/WebKit2InitializeCocoa.mm: Declare default
  profile name for WebKit.
* Source/WebKit/Shared/WebKit2Initialize.cpp: Remove spurious include.
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in: Ditto above.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py: When
  collecting profiles, each invocation will populate /tmp/WebKitPGO.
  Ensure the directory is empty before launching the browser, and copy
  its contents to the diagnostic directory upon exit.
(BenchmarkRunner._run_benchmark):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(run_benchmark_plan): Use a fixed path for writing profiles to.
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner._get_result): Remove the logic that listened
  to the system log for profiling messages and moved profiles out ad hoc.
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig: It static-links
  libWebKitPlatform.a, so when that library is compiled with profiling
  it needs it too. This has likely always been broken, and not been an
  issue since we never attempt to build for profiling with tools.

Canonical link: https://commits.webkit.org/267741@main
